src/gleam/erlang/file.gleam

```diff
@@ -0,0 +1,169 @@
+/// - `Efbig` :: File too large.
+/// - `Eftype` :: Inappropriate file type or format. Usually caused by trying to
+///  set the "sticky bit" on a regular file (not a directory).
+/// - `Eintr` :: Interrupted system call.
+/// - `Einval` :: Invalid argument.
+/// - `Eio` :: I/O error.
+/// - `Eisdir` :: Illegal operation on a directory.
+/// - `Eloop` :: Too many levels of symbolic links.
+/// - `Emfile` :: Too many open files.
+/// - `Emlink` :: Too many links.
+/// - `Emultihop` :: Multihop attempted.
+/// - `Enametoolong` :: Filename too long.
+/// - `Enfile` :: File table overflow
+/// - `Enobufs` :: No buffer space available.
+/// - `Enodev` :: No such device.
+/// - `Enolck` :: No locks available.
+/// - `Enolink` :: Link has been severed.
+/// - `Enoent` :: No such file or directory.
+/// - `Enomem` :: Not enough memory.
+/// - `Enospc` :: No space left on device.
+/// - `Enosr` :: No STREAM resources.
+/// - `Enostr` :: Not a STREAM.
+/// - `Enosys` :: Function not implemented.
+/// - `Enotblk` :: Block device required.
+/// - `Enotdir` :: Not a directory.
+/// - `Enotsup` :: Operation not supported.
+/// - `Enxio` :: No such device or address.
+/// - `Eopnotsupp` :: Operation not supported on socket.
+/// - `Eoverflow` :: Value too large to be stored in data type.
+/// - `Eperm` :: Not owner.
+/// - `Epipe` :: Broken pipe.
+/// - `Erange` :: Result too large.
+/// - `Erofs` :: Read-only file system.
+/// - `Espipe` :: Invalid seek.
+/// - `Esrch` :: No such process.
+/// - `Estale` :: Stale remote file handle.
+/// - `Etxtbsy` :: Text file busy.
+/// - `Exdev` :: Cross-domain link.
+///
+/// ## Erlang Errors
+/// - `Badarg` :: Bad argument; usually the argument is of an invalid type. Due
+///  to Gleam's strong typing, this error should not be encountered when the
+///  functions of this module are called from Gleam code.
+/// - `Terminated` :: File server process is terminated.
+/// - `SystemLimit` :: A system limit was hit, probably not enough ports.
+pub type Reason {
+  Eacces
+  Eagain
+  Ebadf
+  Ebadmsg
+  Ebusy
+  Edeadlk
+  Edeadlock
+  Edquot
+  Eexist
+  Efault
+  Efbig
+  Eftype
+  Eintr
+  Einval
+  Eio
+  Eisdir
+  Eloop
+  Emfile
+  Emlink
+  Emultihop
+  Enametoolong
+  Enfile
+  Enobufs
+  Enodev
+  Enolck
+  Enolink
+  Enoent
+  Enomem
+  Enospc
+  Enosr
+  Enostr
+  Enosys
+  Enotblk
+  Enotdir
+  Enotsup
+  Enxio
+  Eopnotsupp
+  Eoverflow
+  Eperm
+  Epipe
+  Erange
+  Erofs
+  Espipe
+  Esrch
+  Estale
+  Etxtbsy
+  Exdev
+  Bardarg
+  Terminated
+  SystemLimit
+}
+
+/// Read the contents of the given file.
+///
+/// Returns a Result containing the file's contents as a String if the
+/// operation was successful, or Reason if the operation failed.
+///
+/// ## Examples
+///
+///    > read_file("example.txt")
+///    Ok("Hello, World!")
+///
+///    > read_file("does_not_exist.txt")
+///    Error(Enoent)
+///
+pub external fn read_file(filename: String) -> Result(BitString, Reason) =
```

Comment:
How about calling this `read_bits` and having a `read` which we decode to a string? I think to string would be the most common case.

```suggestion
pub external fn read_bits(String) -> Result(BitString, Reason) =
```

Reply:
My biggest question regarding this is what the return type should be. `Result(BitString, Reason)`, as written can represent I/O errors. but cannot represent the situation where a user called `read` on a JPEG. Off-hand, I can think of two possible solutions:
1. Extend `Reason` with a constructor for this specific error, like `NotUTF8` or something.
2. Have `read` return a less descriptive type, something like
   ```gleam
   pub type ReadError {
     IOError
     UTF8Error
   }
   ```
   The former would be returned if any error currently in `Reason` occurred, and the later would be returned if the file's contents are not UTF-8.

Reply:
Given the size of the type already I think it would be more ergonomical to combine the errors into one rather than having two. It's less precise but more approachable.